src/IMDS.jl

```
@@ -0,0 +1,162 @@
+        # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instance-metadata-returns
+        if e isa ConnectError || e isa StatusError && e.status == 403
+            throw(IMDSUnavailable())
+
+        #! format: off
+        # Return the status exception when `status_exception=false`. We must always cause
+        # `HTTP.request` to throw status errors for our `IMDSUnavailable` check.
+        #! format: on
+        elseif !status_exception && e isa StatusError
+            e.response
+        else
+            rethrow()
+        end
+    end
+
+    return response
+end
+
+"""
+    get([session::Session], path::AbstractString) -> Union{String, Nothing}
+
+Retrieve the AWS instance metadata from the provided HTTP `path`. If no instance metadata is
+available (due to the instance metadata service being disabled or not being run from within
+an EC2 instance) then `nothing` will be returned. For details on available metadata see the
+official AWS documentation on:
+["Instance metadata and user data"](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html).
+
+# Arguments
+- `session` (optional): The IMDS `Session` used to store the IMDSv2 token.
+- `path`: The HTTP path to used to specify the metadata to return.
+"""
+function get(session::Session, path::AbstractString)
+    response = try
+        request(session, "GET", path)
+    catch e
+        if e isa IMDSUnavailable
+            nothing
+        else
+            rethrow()
+        end
+    end
+
+    return !isnothing(response) ? String(response.body) : nothing
+end
+
+"""
+    region([session::Session]) -> Union{String, Nothing}
+
+Determine the AWS region of the machine executing this code if running inside of an EC2
+instance, otherwise `nothing` is returned.
+
+# Arguments
+- `session` (optional): The IMDS `Session` used to store the IMDSv2 token.
+"""
+region(session::Session) = get(session, "/latest/meta-data/placement/region")
```

Replaces this deprecated code:
```julia
ec2_instance_region() =
    try
        ec2_instance_metadata("/latest/meta-data/placement/region")
    catch
        nothing
    end
```

The new code no longer suppresses any exceptions that occur. I believe the reason we originally did this was due to instance metadata failures we saw when running tests against GHA CI as there is a webservice running on the CI instance at the expected instance metadata IP address that isn't a instance metadata service. I've since updated the tests to correct this problem.

However, this is still technically a breaking code change. I'm not concerned about but if a review is we can always put the guard rails on the deprecation.

---

The try/catch portion was added in https://github.com/JuliaCloud/AWS.jl/pull/580 which seems to be an work around for failing testcases